mypyc/codegen/emitfunc.py

```diff
@@ -216,17 +236,20 @@ def visit_branch(self, op: Branch) -> None:
 
         if false is self.next_block:
             if op.traceback_entry is None:
-                self.emit_line(f"if ({cond}) goto {self.label(true)};")
+                if true is not self.next_block:
+                    self.emit_line(f"if ({cond}) goto {self.label(true)};")
```

Comment:
I just tried undoing this hunk, and Mypyc won't compile without it:

```
      build/__native_26051150be1624671934.c:7975:26: error: use of undeclared label 'CPyL11'
          if (!cpy_r_r28) goto CPyL11;
                               ^
      build/__native_26051150be1624671934.c:85644:26: error: use of undeclared label 'CPyL21'
          if (!cpy_r_r92) goto CPyL21;
                               ^
      build/__native_26051150be1624671934.c:343800:26: error: use of undeclared label 'CPyL18'
          if (!cpy_r_r56) goto CPyL18;
                               ^
      build/__native_26051150be1624671934.c:374890:26: error: use of undeclared label 'CPyL14'
          if (!cpy_r_r53) goto CPyL14;
                               ^
      build/__native_26051150be1624671934.c:411655:26: error: use of undeclared label 'CPyL18'
          if (!cpy_r_r63) goto CPyL18;
                               ^
      build/__native_26051150be1624671934.c:708426:26: error: use of undeclared label 'CPyL10'
          if (!cpy_r_r32) goto CPyL10;
                               ^
...
```

I assume that there are places in the code where a `Branch` is created/updated in such a way that the `true` and `false` branches are the same. I agree that this shouldn't be possible, but it is not enforced. Perhaps we should use an `assert` to ensure that the `true` and `false` branches cannot be the same?

Reply:
I added an `assert` just to check, it seems to be getting stuck on `if TYPE_CHECKING` branches (might just be a coincidence).

Debug/asserts I added:

```diff
     def visit_branch(self, op: Branch) -> None:
+        true, false = op.true, op.false
+
+        print(op.value, op.rare)
+        print(self.module_name, self.source_path, op.line)
+
+        assert true is not false
```

Output:

```
...

      <mypyc.ir.ops.CallC object at 0x7f7b49e684c0> False
      mypy.argmap mypy/argmap.py 20
      <mypyc.ir.ops.Unbox object at 0x7f7b49e684f0> False
      mypy.argmap mypy/argmap.py 20
      <mypyc.ir.ops.Unbox object at 0x7f7b49e684f0> False
      mypy.argmap mypy/argmap.py -1
      Traceback (most recent call last):
        File "mypy/.venv/lib/python3.10/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 363, in <module>
          main()
        File "mypy/.venv/lib/python3.10/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 345, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
        File "mypy/.venv/lib/python3.10/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 144, in get_requires_for_build_editable
          return hook(config_settings)
        File "/tmp/pip-build-env-1msl1q9d/overlay/lib/python3.10/site-packages/setuptools/build_meta.py", line 450, in get_requires_for_build_editable
          return self.get_requires_for_build_wheel(config_settings)
        File "/tmp/pip-build-env-1msl1q9d/overlay/lib/python3.10/site-packages/setuptools/build_meta.py", line 341, in get_requires_for_build_wheel
          return self._get_build_requires(config_settings, requirements=['wheel'])
        File "/tmp/pip-build-env-1msl1q9d/overlay/lib/python3.10/site-packages/setuptools/build_meta.py", line 323, in _get_build_requires
          self.run_setup()
        File "/tmp/pip-build-env-1msl1q9d/overlay/lib/python3.10/site-packages/setuptools/build_meta.py", line 338, in run_setup
          exec(code, locals())
        File "<string>", line 169, in <module>
        File "mypy/mypyc/build.py", line 511, in mypycify
          groups, group_cfilenames = mypyc_build(
        File "mypy/mypyc/build.py", line 421, in mypyc_build
          group_cfiles, ops_text = generate_c(
        File "mypy/mypyc/build.py", line 229, in generate_c
          modules, ctext = emitmodule.compile_modules_to_c(
        File "mypy/mypyc/codegen/emitmodule.py", line 426, in compile_modules_to_c
          ctext = compile_ir_to_c(groups, modules, result, mapper, compiler_options)
        File "mypy/mypyc/codegen/emitmodule.py", line 310, in compile_ir_to_c
          ctext[group_name] = generator.generate_c_for_modules()
        File "mypy/mypyc/codegen/emitmodule.py", line 550, in generate_c_for_modules
          generate_native_function(fn, emitter, self.source_paths[module_name], module_name)
        File "mypy/mypyc/codegen/emitfunc.py", line 164, in generate_native_function
          ops[visitor.op_index].accept(visitor)
        File "mypy/mypyc/ir/ops.py", line 410, in accept
          return visitor.visit_branch(self)
        File "mypy/mypyc/codegen/emitfunc.py", line 204, in visit_branch
          assert true is not false
      AssertionError
```

And this is what is on line 20 of `mypy/argmap.py`:

https://github.com/python/mypy/blob/25047b87ffb961b787481af39ce428ef52fe08bd/mypy/argmap.py#L20-L21

Reply:
Hmm yeah. I took a look into this and it's nontrivial to fix. This is happening because those imports are marked as internal and aren't processed, so the if statement's body block contains only a Goto op to the next block. Then, an optimization pass in the dataflow analysis deletes the body block, updating the branch op to directly jump to the next block (and consequently setting `true` and `false` to same block).

Let's leave this for a separate issue. Thanks for digging into this!

Reply:
@ichard26 Was an issue filed? I'm getting a `get_requires_for_build_editable` failure with mypy-1.4.0

https://github.com/common-workflow-language/cwltool/actions/runs/5353323117/jobs/9730454728#step:8:114